File: ABC/351-400/397/c.py
def main():
  n = int(input())
  a = list(map(int,input().split()))
  fr = [0]*n
  bk = [0]*n
  st = set()
  for i in range(n):
    st.add(a[i])
    fr[i] = len(st)

  st = set()
  for i in range(n-1,-1,-1):
    st.add(a[i])
    bk[i] = len(st)
  ans = 0
  for i in range(n-1):
     ans = max(ans,fr[i]+bk[i+1])

  print(ans)
if __name__ == '__main__':
  main()

File: ABC/351-400/397/c.cpp
#include <iostream>
#include<set>
#include<vector>
using namespace std;

int main(){
  int n;
  cin >> n;
  vector<int> a(n);
  for(int i=0;i<n;i++) cin >> a[i];
  vector<int> f(n),b(n);
  set<int> st1;
  for(int i=0;i<n;i++){
    st1.insert(a[i]);
    f[i] = st1.size();
  }
  set<int> st2;
  for(int i=n-1;i>=0;i--){
    st2.insert(a[i]);
    b[i] = st2.size();
  }
  int ans = 0;
  for(int i=0;i<n-1;i++){
    ans = max(ans,f[i]+b[i+1]);
  }
  cout << ans << endl;
  return 0;
}

